src/FacetedSearchContainer.js

```diff
@@ -0,0 +1,153 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+import _ from 'lodash'
+
+import FilterSidebar from './FilterSidebar'
+import FilterList from './FilterList'
+
+import {ResultPropTypes} from './ResultPropTypes'
+
+class FacetedSearchContainer extends React.Component {
+  constructor(props) {
+    super(props)
+
+    this.state = {
+      facets:
+        _(props.results)
+          .flatMap(`facets`)
+          .compact()   // lodash will emit undefined if `facets` doesn’t exist :/
+          .uniqWith(_.isEqual)
+          .map((facet) => ({...facet, disabled: false}))
+          .value(),
+      selectedFacets: {}  // TODO (?) Build initial state of checked filters from props if wrapped in React Router
+    }
+
+    this._handleChange = this._handleChange.bind(this)
+  }
+
+  _filterResults(facets) {
+    return this.props.results.filter((result) =>
+      // Leave the results that, for every group, have at least one matching property within that group
+      // Note: every returns true for empty arrays
+      Object.entries(facets)
+        .every(
+          ([facetGroup, facetValues]) =>
+            result.facets.some(
+              (resultFacet) =>
+                facetGroup === resultFacet.group && facetValues.map(f => f.value).includes(resultFacet.value))))
+  }
+
+  _hasNoResults(selectedFacets, additionalFacet) {
+    const mergedFacets =
+      _.mergeWith(
+        _.defaultsDeep({}, selectedFacets),
+        {[additionalFacet.group]: additionalFacet},
+        (objValue, srcValue) => _.uniq((objValue || []).concat(srcValue))
+      )
+
+    return this._filterResults(mergedFacets).length === 0
+  }
+
+  _disableEnabledFacetsWithNoResults(selectedFacets, facetGroup) {
+    return _.cloneDeep(this.state.facets)
+      .map((facet) => ({
+        ...facet,
+        disabled:
+          facet.group !== facetGroup ?
+            facet.disabled ? true : this._hasNoResults(selectedFacets, facet) :
+            facet.disabled
+      }))
+  }
+
+  _enableDisabledFacetsWithResults(selectedFacets, facetGroup) {
+    return _.cloneDeep(this.state.facets)
+      .map((facet) => ({
+        ...facet,
+        disabled:
+          facet.group !== facetGroup ?
+            facet.disabled ? this._hasNoResults(selectedFacets, facet) : false :
+            facet.disabled
+      }))
+  }
+
+  _handleChange(facetGroup, selectedFacetsInGroup) {
+    const _selectedFacets = _.defaultsDeep({}, this.state.selectedFacets)
+    _selectedFacets[facetGroup] = selectedFacetsInGroup
+
+    const nextSelectedFacets =
+       Object.keys(_selectedFacets)
+         .filter((key) => _selectedFacets[key] && _selectedFacets[key].length > 0)
+         .reduce((acc, key) => {
+           acc[key] = _selectedFacets[key]
```

Comment:
I'm guessing an `acc` is an accumulator?

Reply:
Yes it is! I accept suggestions though.